contracts/IdleTokenFungible.sol

```diff
@@ -0,0 +1,759 @@
+   */
+  function redeemIdleToken(uint256 _amount)
+    external
+    returns (uint256) {
+      return _redeemIdleToken(_amount);
+  }
+
+  /**
+   * Here we calc the pool share one can withdraw given the amount of IdleToken they want to burn
+   *
+   * @param _amount : amount of IdleTokens to be burned
+   * @return redeemedTokens : amount of underlying tokens redeemed
+   */
+  function _redeemIdleToken(uint256 _amount)
+    internal nonReentrant
+    returns (uint256 redeemedTokens) {
+      _checkMintRedeemSameTx();
+      _updateFeeInfo();
+      if (_amount != 0) {
+        uint256 price = _tokenPrice();
+        uint256 valueToRedeem = _amount * price / ONE_18;
+        uint256 balanceUnderlying = _contractBalanceOf(token);
+
+        if (valueToRedeem > balanceUnderlying) {
+          redeemedTokens = _redeemHelper(_amount, balanceUnderlying);
+        } else {
+          redeemedTokens = valueToRedeem;
+        }
+        // update lastNAV
+        lastNAV -= redeemedTokens;
+        // burn idleTokens
+        _burn(msg.sender, _amount);
+        // send underlying minus fee to msg.sender
+        _transferTokens(token, msg.sender, redeemedTokens);
+      }
+  }
+
+  function _redeemHelper(uint256 _amount, uint256 _balanceUnderlying) private returns (uint256 redeemedTokens) {
+    address currToken;
+    uint256 idleSupply = totalSupply();
+    address[] memory _allAvailableTokens = allAvailableTokens;
+
+    for (uint256 i = 0; i < _allAvailableTokens.length; i++) {
+      currToken = _allAvailableTokens[i];
+      redeemedTokens += _redeemProtocolTokens(
+        currToken,
+        // _amount * protocolPoolBalance / idleSupply
+        _amount * _contractBalanceOf(currToken) / idleSupply // amount to redeem
+      );
+    }
+    // and get a portion of the eventual unlent balance
+    redeemedTokens += _amount * _balanceUnderlying / idleSupply;
+  }
+
+  /**
+   * Dynamic allocate all the pool across different lending protocols if needed,
+   * rebalance without params
+   *
+   * NOTE: this method can be paused
+   *
+   * @return : whether has rebalanced or not
+   */
+  function rebalance() external returns (bool) {
+    return _rebalance();
+  }
+
+  // internal
+  /**
+   * Get current idleToken price based on net asset value and totalSupply
+   *
+   * @return price value of 1 idleToken in underlying
+   */
+  function _tokenPrice() internal view returns (uint256 price) {
+    uint256 totSupply = totalSupply();
+    if (totSupply == 0) {
+      return 10**(tokenDecimals);
+    }
+
+    address currToken;
+    uint256 totNav = _contractBalanceOf(token) * ONE_18; // eventual underlying unlent balance
+    address[] memory _allAvailableTokens = allAvailableTokens;
+    uint256 _len = _allAvailableTokens.length;
+    for (uint256 i = 0; i < _len; ) {
+      currToken = _allAvailableTokens[i];
+      // NAV = price * poolSupply
+      totNav += _getPriceInToken(protocolWrappers[currToken]) * _contractBalanceOf(currToken);
```

Comment:
we don't need to divide the value with 1e18, because priceInUnderlying * balanceInUnderlying => value in underlying, right

Reply:
If we use USDC as example:
at the beginning of the method we have ` uint256 totNav = _contractBalanceOf(token) * ONE_18` so totNAV is something like 1e6 * 1e18

Then at line 444 we add 2 values that are in the order of 1e6 for usdc so totNav is still in the format 1e6 * 1e18

last line we do `price = (totNav - unclaimedFees) / totSupply;` -> price = 1e6 * 1e18 / 1e18 = 1e6 (idleToken has always 18 decimals so totSupply is in the format of 1e18).

So the price at the end has 6 decimals ie the same of the underlying asset.

Reply:
got it.